### PR TITLE
Dataset & Visualization callback

### PR DESCRIPTION
**Pull request by BloodAxe**
New dataset class for COCO/CrowdPose
Visualization callback for pose estimation
Dataset params for COCO & Crowdpose datasets

### CODE REVIEW

**Review by Louis-Dupont — APPROVED** [review]
LGTM

**Review by ofrimasad — COMMENTED** [review]
looks good.
Maybe we can make the yamls smaller (looks like redundant overrides) 
Some comment inline

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        return len(self.ids)
+
+    def load_sample(self, index: int) -> PoseEstimationSample:
+        """
+        Read a sample from the disk and return a PoseEstimationSample
+        :param index: Sample index
+        :return:      Returns an instance of PoseEstimationSample that holds complete sample (image and annotations)
+        """
+        img_id = self.ids[index]
+        image_info = self.coco.loadImgs(img_id)[0]
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py:132]: @ofrimasad honestly, I haven't noticed any slowness in working with current implementation. Dataset instantiation certainly takes un-noticeable amount of time, nothing similar to what we have in detection.
And overall training is quite fast - the S model takes around 4 minutes on 8 gpus to process so I think we are good in this sense.

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        """
+
+        :param data_dir:                     Root directory of the COCO dataset
+        :param images_dir:                   Path suffix to the images directory inside the dataset_root
+        :param json_file:                    Path suffix to the json file inside the dataset_root
+        :param include_empty_samples:        If True, images without any annotations will be included in the dataset.
+                                             Otherwise, they will be filtered out.
+        :param transforms:                   Transforms to be applied to the image & keypoints
+        :param edge_links:                   Edge links between joints
+        :param edge_colors:                  Color of the edge links. If None, the color will be generated randomly.
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py:63]: We need to have these parameters because they are dataset-specific and stored as part of preprocessing parameters in the checkpoint metadata to enable .predict() function to work (Predict needs to know how to draw the pose with needs edge_links and edge_colors)

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  json_file: annotations/person_keypoints_train2017.json
+
+  edge_links: ${dataset_params.edge_links}
+  edge_colors: ${dataset_params.edge_colors}
+  keypoint_colors: ${dataset_params.keypoint_colors}
+
+  include_empty_samples: True
+  crowd_annotations_action: mask_as_normal
+
+  transforms:
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml:21]: Yes, default `crowd_annotations_action` is `no_action` and `include_empty_samples` has no default.

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml [hunk 4]
```diff
@@ -0,0 +1,109 @@
+defaults:
+  - coco_pose_common_dataset_params
+  - _self_
+
+# This is a shortcut parameter to set size of training & validation images.
+image_size: 640
+dataset_params_suffix: "${dataset_params.image_size}"
+
+train_dataset_params:
+  data_dir: /data/coco # root path to coco data
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml:10]: `coco_pose_estimation_yolo_nas_dataset_params.yaml` inherits only from `coco_pose_common_dataset_params.yaml` 

The `coco_pose_common_dataset_params.yaml`  contains only `oks_sigmas`, `flip_indexes`, `edge_links` and some other fields but no `train_dataset_params`. So there it is not overriding `data_dir` and other fields.  

This is done intentionally to make this file less dependent on existing `coco_pose_estimation_common_dataset_params.yaml` which we still want to keep unchanged for compatibility reasons. But `coco_pose_estimation_common_dataset_params.yaml` is tailored for old dataset classes

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml:1]: I think it would be a bit confusing since we explicitly use dataset class & collate function in this YAM from YoloNAS Pose:

```
train_dataloader_params:
  dataset: COCOPoseEstimationDataset
  ...
  collate_fn: YoloNASPoseCollateFN

```

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                                             AP metric on validation greatly. This option allows to remove these duplicates.
+                                             However, it is disabled by default to preserve backward compatibility with COCO evaluation.
+        :param crowd_annotations_action:     Action to take for annotations with iscrowd=1. Can be one of the following:
+                                             "drop_sample" - Samples with crowd annotations will be dropped from the dataset.
+                                             "drop_annotation" - Crowd annotations will be dropped from the dataset.
+                                             "mask_as_normal" - These annotations will be treated as normal (non-crowd) annotations.
+                                             "no_action" - No action will be taken for crowd annotations.
+        """
+
+        crowd_annotations_action = CrowdAnnotationActionEnum(crowd_annotations_action)
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py]: Fixed in 0eeb8f2b94bbbdcdef9acd67d8a6dd6e00dd4f46, thanks!

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+        :param data_dir:                     Root directory of the COCO dataset
+        :param images_dir:                   Path suffix to the images directory inside the dataset_root
+        :param json_file:                    Path suffix to the json file inside the dataset_root
+        :param include_empty_samples:        If True, images without any annotations will be included in the dataset.
+                                             Otherwise, they will be filtered out.
+        :param transforms:                   Transforms to be applied to the image & keypoints
+        :param edge_links:                   Edge links between joints
+        :param edge_colors:                  Color of the edge links. If None, the color will be generated randomly.
+        :param keypoint_colors:              Color of the keypoints. If None, the color will be generated randomly.
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py:68]: See https://github.com/Deci-AI/super-gradients/pull/1520#discussion_r1358013984

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        :param remove_duplicate_annotations: If True will remove duplicate instances from the dataset.
+                                             It is known that COCO dataset has some duplicate annotations that and that affects the
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py]: Fixed in ba6a74536d363bf84b46d2bfcd49355e3c524a69

File: src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  images_dir: images/train2017
+  json_file: annotations/person_keypoints_train2017.json
+
+  edge_links: ${dataset_params.edge_links}
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/recipes/dataset_params/coco_pose_estimation_yolo_nas_dataset_params.yaml:14]: Probably duplicate of https://github.com/Deci-AI/super-gradients/pull/1520#discussion_r1356506156 right?

File: src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            coco = remove_duplicate_annotations_fn(coco)
+
+        if crowd_annotations_action == CrowdAnnotationActionEnum.DROP_SAMPLE:
+            coco = remove_samples_with_crowd_annotations(coco)
+        elif crowd_annotations_action == CrowdAnnotationActionEnum.DROP_ANNOTATION:
+            coco = remove_crowd_annotations(coco)
+
+        if len(coco.dataset["categories"]) != 1:
+            raise ValueError("Dataset must contain exactly one category")
+        joints = coco.dataset["categories"][0]["keypoints"]
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/datasets/pose_estimation_datasets/coco_pose_estimation_dataset.py:97]: Fixed in 0eeb8f2b94bbbdcdef9acd67d8a6dd6e00dd4f46

**Review by Louis-Dupont — APPROVED** [review]
LGTM
